http://gcc.gnu.org/PR32500

2007-07-04  Richard Guenther  <rguenther@suse.de>

	PR tree-optimization/32500
	* tree-ssa-loop-niter.c (infer_loop_bounds_from_undefined):
	Only use basic blocks that are always executed to infer loop bounds.

--- branches/gcc-4_2-branch/gcc/tree-ssa-loop-niter.c	2007/07/04 11:44:58	126314
+++ branches/gcc-4_2-branch/gcc/tree-ssa-loop-niter.c	2007/07/04 12:38:23	126315
@@ -1747,6 +1747,12 @@
     {
       bb = bbs[i];
 
+      /* If BB is not executed in each iteration of the loop, we cannot
+	 use the operations in it to infer reliable upper bound on the
+	 # of iterations of the loop.  */
+      if (!dominated_by_p (CDI_DOMINATORS, loop->latch, bb))
+	continue;
+
       for (bsi = bsi_start (bb); !bsi_end_p (bsi); bsi_next (&bsi))
         {
 	  tree stmt = bsi_stmt (bsi);
